frontend/dialogs/OBSBasicSourceSelect.cpp

```diff
@@ -232,59 +175,473 @@ bool AddNew(QWidget *parent, const char *id, const char *name, const bool visibl
 
-			obs_scene_t *scene = obs_get_scene_by_name(scene_name);
-			OBSSceneItem item;
-			auto cb = [](obs_scene_t *, obs_sceneitem_t *sceneitem, void *data) {
-				OBSSceneItem &last = *static_cast<OBSSceneItem *>(data);
-				last = sceneitem;
-				return true;
-			};
-			obs_scene_enum_items(scene, cb, &item);
+	return true;
+}
 
-			obs_sceneitem_remove(item);
-			obs_scene_release(scene);
-		};
+void OBSBasicSourceSelect::OBSSourceAdded(void *data, calldata_t *calldata)
+{
+	OBSBasicSourceSelect *window = static_cast<OBSBasicSourceSelect *>(data);
+	obs_source_t *source = (obs_source_t *)calldata_ptr(calldata, "source");
 
-		auto redo = [scene_name, main, source_name, visible](const std::string &) {
-			obs_source_t *scene_source = obs_get_source_by_name(scene_name);
-			main->SetCurrentScene(scene_source, true);
-			obs_source_release(scene_source);
-			AddExisting(QT_TO_UTF8(source_name), visible, false, nullptr, nullptr, nullptr, nullptr);
-		};
+	QMetaObject::invokeMethod(window, "SourceAdded", Q_ARG(OBSSource, source));
+}
+
+void OBSBasicSourceSelect::getSourceTypes()
+{
+	OBSBasic *main = reinterpret_cast<OBSBasic *>(App()->GetMainWindow());
+
+	const char *unversioned_type;
+	const char *type;
+
+	size_t idx = 0;
+
+	while (obs_enum_input_types2(idx++, &type, &unversioned_type)) {
+		const char *name = obs_source_get_display_name(type);
+		uint32_t caps = obs_get_source_output_flags(type);
+
+		if ((caps & OBS_SOURCE_CAP_DISABLED) != 0)
+			continue;
+
+		QListWidgetItem *newItem = new QListWidgetItem();
+		newItem->setData(Qt::DisplayRole, name);
+		newItem->setData(UNVERSIONED_ID_ROLE, unversioned_type);
 
-		undo_s.add_action(QTStr("Undo.Add").arg(source_name), undo, redo, "", "");
+		QIcon icon;
+		icon = main->GetSourceIcon(type);
+		newItem->setIcon(icon);
+
+		ui->sourceTypeList->addItem(newItem);
+	}
+
+	QListWidgetItem *newItem = new QListWidgetItem();
+	newItem->setData(Qt::DisplayRole, Str("Basic.Scene"));
+	newItem->setData(UNVERSIONED_ID_ROLE, "scene");
+
+	QIcon icon;
+	icon = main->GetSceneIcon();
+	newItem->setIcon(icon);
+
+	ui->sourceTypeList->addItem(newItem);
+
+	ui->sourceTypeList->sortItems();
+
+	QListWidgetItem *allSources = new QListWidgetItem();
+	allSources->setData(Qt::DisplayRole, Str("Basic.SourceSelect.Recent"));
+	allSources->setData(UNVERSIONED_ID_ROLE, QVariant());
+	ui->sourceTypeList->insertItem(0, allSources);
+
+	ui->sourceTypeList->setCurrentItem(allSources);
+	ui->sourceTypeList->setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Preferred);
+
+	connect(ui->sourceTypeList, &QListWidget::currentItemChanged, this, &OBSBasicSourceSelect::sourceTypeSelected);
+}
+
+void OBSBasicSourceSelect::setSelectedSourceType(QListWidgetItem *item)
+{
+	setSelectedSource(nullptr);
+	QLayout *layout = ui->existingListFrame->flowLayout();
+
+	// Clear existing buttons when switching types
+	QLayoutItem *child = nullptr;
+	while ((child = layout->takeAt(0)) != nullptr) {
+		delete child->widget();
+		delete child;
```

Comment:
Bad practice, but I guess that's the way we have to do it, because even though while the parent widget is the "owner" of the child elements (and we rely on the parent to handle the lifetime of those elements by default), we interfere in this by prematurely ending the lifetime of these children.

I'm surprised that `QWidget` does not have a "removeChildAt` method or something similar, as that would retain the ownership relationship (it's still the parent ending the lifetime) but allows updating the widget hierarchy more gracefully.